enhancements/network/on-prem-mutable-vips.md

```diff
@@ -0,0 +1,199 @@
+
+### User Stories
+
+* As a deployer of a dual-stack OpenShift cluster, I want to access the API using both IPv4 and IPv6.
+
+* As a deployer of a dual-stack OpenShift cluster, I want to access the Ingress using both IPv4 and IPv6.
+
+### Goals
+
+Allow adding a second pair of virtual IPs to an already installed dual-stack cluster.
+
+### Non-Goals
+
+* Modifying existing virtual IP configuration.
+
+* Modifying virtual IP configuration after second pair of VIPs has been added. Right now we only want to cover "create" operation. Rest of CRUD operations will be covered in the future depending on the needs.
+
+* Configuration of any VIPs beyond a second pair for the second IP stack. MetalLB is a better solution for creating arbitrary loadbalancers.
+
+## Proposal
+
+### Workflow Description
+
+The proposed worklflow after implementing the feature would look as described below
+
+1. Administrator of a dual-stack cluster with single-stack VIPs wants to add a second pair for the second IP stack configured.
+
+1. Administrator edits the Infrastructure CR named `cluster` by changing the `spec.platformSpec.[*].apiServerInternalIPs` and `spec.platformSpec.[*].ingressIPs` fields.
+
+1. Cluster Network Operator picks the modification of the object and compares values with `spec.platformStatus.[*].apiServerInternalIPs` and `spec.platformStatus.[*].ingressIPs`.
+
+1. After validating that the requested change is valid (i.e. conforms with the goals and non-goals), the change is propagated down to the keepalived template and configuration file.
+
+1. After keepalived configuration is changed, the service is restarted or reloaded to apply the changes.
+
+### API Extensions
+
+New fields will have to be added to the platform spec section for [baremetal](https://github.com/openshift/api/blob/938af62eda38e539488d6193e7af292e47a09a5e/config/v1/types_infrastructure.go#L694), [vsphere](https://github.com/openshift/api/blob/938af62eda38e539488d6193e7af292e47a09a5e/config/v1/types_infrastructure.go#L1089) and [OpenStack](https://github.com/openshift/api/blob/938af62eda38e539488d6193e7af292e47a09a5e/config/v1/types_infrastructure.go#L772). In the first implementation only baremetal will be covered.
+
+### Implementation Details/Notes/Constraints
+
+Because of the expertise and the team owning this feature, we think the best place for implementing the logic is Cluster Network Operator. CNO however is not an owner of the `cluster` Infrastructure CR what at first sight may seem like a challenge.
+
+Kubernetes API provide a solution for this and we can leverage Watches (with Predicate for optimization) to allow a controller inside CNO to watch an object it does not own. This in fact already happens inside CNO in the [infrastructureconfig controller](https://github.com/openshift/cluster-network-operator/blob/2005bcd8c93de5bffc05c9c943b51386007f6b9a/pkg/controller/infrastructureconfig/infrastructureconfig_controller.go#L47).
```

Comment:
Just a note for anyone else reading this who isn't familiar with the history: This logic was added to CNO as part of the initial dual stack VIPs implementation. We had discussions with multiple teams about where that logic should live and CNO was the eventual winner.

Reply:
Added this directly to the doc